burn-autodiff/src/checkpoint/base.rs

```diff
@@ -0,0 +1,172 @@
+use std::collections::HashMap;
+
+use crate::graph::{NodeID, NodeRef};
+
+use super::state::State;
+
+/// Definition of the forward function of a node, called during retropropagation only.
+/// This is different from the normal forward function because it reads and writes from
+/// the [InnerStates] map instead of having a clear function signature.
+pub(crate) trait RetroForward {
+    fn forward(&self, states: &mut InnerStates);
+}
+
+#[derive(new, Default)]
+/// Links [NodeID]s to their corresponding [RetroForward]
+pub(crate) struct RetroForwards {
+    map: HashMap<NodeID, Box<dyn RetroForward>>,
+}
+
+impl RetroForwards {
+    /// Executes the [RetroForward] for a given [NodeID] if the node's
+    /// [State] is [State::Recompute], otherwise does nothing.
+    pub fn forward(&self, node_id: &NodeID, inner_states: &mut InnerStates) {
+        if let State::Recompute { n_required: _ } = inner_states.get_ref(node_id).unwrap() {
+            self.map.get(&node_id).unwrap().forward(inner_states);
+        }
+    }
+
+    /// Associates a [RetroForward] to its [NodeID]
+    pub fn insert(&mut self, node_id: NodeID, retro_forward: Box<dyn RetroForward>) {
+        self.map.insert(node_id, retro_forward);
+    }
+}
+
+#[derive(new, Default)]
+/// Links [NodeID]s to their current [State]
+pub(crate) struct InnerStates {
+    map: HashMap<NodeID, State>,
+}
+
+impl InnerStates {
+    /// Returns the output in the [State] of the given [NodeID],
+    /// and decrements the number of times this state is required.
+    /// This function always gives ownership of the output, but will clone it if needed for further uses.
+    pub fn get_owned_and_downcasted<T>(&mut self, node_id: &NodeID) -> T
+    where
+        T: Clone + Send + Sync + 'static,
+    {
+        // Fetch the state and decrement its number of required
+        let state = self.map.remove(node_id).unwrap();
+        let remaining_n_required = state.n_required() - 1;
+
+        // Downcast the state to whatever it is supposed to be
+        let downcasted = state
+            .get_state_content()
+            .downcast_ref::<T>()
+            .unwrap()
+            .clone();
```

Comment:
You don't want to clone the state content here, you want to take ownweship of it. 

```rust

   // Fetch the state and decrement its number of required
        let state = self.map.remove(node_id).unwrap();
        let remaining_n_required = state.n_required() - 1;

        // Downcast the state to whatever it is supposed to be
        // If still needed after giving ownership, we copy it back to the hashmap
        if remaining_n_required > 0 {
            let new_stored_state = match state {
                State::Recompute { n_required: _ } => State::Recompute {
                    n_required: remaining_n_required,
                },
                State::Computed {
                    state_content,
                    n_required: _,
                } => State::Computed {
                    state_content,
                    n_required: remaining_n_required,
                },
            };

            let downcasted = new_stored_state
                .to_state_content()
                .downcast_ref::<T>()
                .unwrap()
                .clone();

            self.insert(node_id.clone(), new_stored_state);

            downcasted
        } else {
            let downcasted = state.into_state_content().downcast::<T>().unwrap();

            *downcasted
        }
 ```

You need `to` and `into` variant of `state_content()`